### ISSUE TITLE
Ensure custom Function are correct in double backward setting

### ISSUE THREAD

**Issue by albanD** [pytorch/pytorch#76185]
### Problem statement
Today, if intermediary Tensors are saved during the forward and used during the backward, the double backward will be silently wrong because their contribution will be ignored.

A small repro is:
```python
import torch
from torch.autograd import Function, gradcheck, gradgradcheck


class MyFn(Function):
    @staticmethod
    def forward(ctx, inp):
        temp = inp * 2
        out = temp * 3
        ctx.temp = temp
        ctx.inp = inp
        return out

    @staticmethod
    def backward(ctx, gO):
        return gO * torch.full_like(gO, 3) * ctx.temp / ctx.inp


inp = torch.rand(10, dtype=torch.float64, requires_grad=True)

gradcheck(MyFn.apply, inp)
gradgradcheck(MyFn.apply, inp)
```

### Proposed solution
Note that I am not sure this is the best way to go about this, we might want to do something else in practice?

Option 1) Add a `ErrorNode` as the grad_fn for everything that gets saved on the ctx and via save_for_backward if it does not already have a grad_fn.
Since the backward runs in no grad_mode when you're not doing double backward, it will just be ignored for most users. But in the double backward case, it will be attached to the created graph. And so if the user tries to backward through that particular Tensor, we can provide a nice error message.
If this is a false positive for some users (they want to ignore that intermediary from the double backward computation), then they can just "detach()" it to recover the previous behavior.

Option 2) Add these error nodes on the fly during the backward function only if grad_mode is enabled. This avoid mutating the object when the user does `ctx.foo = foo` which is better? And only the accessed one is changed? Unfortunately, we cannot do this out of place as some users rely on the fact that `ctx.foo is foo`.

Option 3) ??

**Comment by ezyang** [comment]
option 1 seems like the right way to me

**Comment by soulitzer** [comment]
From offline discussion @albanD mentions that does not have `grad_fn` is in the c++ sense, so accumulate grad should also  count as grad_fn

**Comment by caiopetruccirosa** [comment]
Is this issue still open? If it is, I would like to work on it. :)
